src/function/cast_rules.cpp

```diff
@@ -437,19 +440,87 @@ int64_t CastRules::ImplicitCast(const LogicalType &from, const LogicalType &to)
 		}
 		return cost;
 	}
+	if (from.id() == LogicalTypeId::STRUCT && to.id() == LogicalTypeId::STRUCT) {
+		if (to.AuxInfo() == nullptr) {
+			// If this struct is not fully resolved, we'll leave it to the actual cast logic to handle it.
+			return 0;
+		}
+
+		auto &source_children = StructType::GetChildTypes(from);
+		auto &target_children = StructType::GetChildTypes(to);
+
+		if (source_children.size() != target_children.size()) {
+			// different number of children: not possible
+			return -1;
+		}
+
+		auto target_is_unnamed = StructType::IsUnnamed(to);
+		auto source_is_unnamed = StructType::IsUnnamed(from);
+		auto named_struct_cast = !source_is_unnamed && !target_is_unnamed;
+
+		if (named_struct_cast) {
+			int64_t cost = -1;
+
+			// Collect the target members in a map for easy lookup
+			case_insensitive_map_t<idx_t> target_members;
+			for (idx_t target_idx = 0; target_idx < target_children.size(); target_idx++) {
+				auto &target_name = target_children[target_idx].first;
+				if (target_members.find(target_name) != target_members.end()) {
+					// duplicate name in target struct
+					return -1;
+				}
+				target_members[target_name] = target_idx;
+			}
+
+			for (idx_t source_idx = 0; source_idx < source_children.size(); source_idx++) {
+				auto &source_child = source_children[source_idx];
+				auto entry = target_members.find(source_child.first);
+				if (entry == target_members.end()) {
+					// element in source struct was not found in target struct
+					return -1;
+				}
+				auto target_idx = entry->second;
+				target_members.erase(entry);
+				auto child_cost = ImplicitCast(source_child.second, target_children[target_idx].second);
```

Comment:
We should probably return `-1` if any of the children is not implicitly castable, no?